pyinstrument/renderers/console.py

```diff
@@ -137,11 +147,64 @@ def render_frame(self, frame: Frame, indent: str = "", child_indent: str = "") -
 
         return result
 
-    def frame_proportion_of_total_time(self, frame: Frame):
-        return frame.time / self.root_frame.time
+    def render_frame_flat(self, frame: Frame) -> str:
+        @dataclass(frozen=True, eq=True)
+        class FrameDesc:
+            code_position_short: Optional[str]
+            function: str
+
+
+        def walk(frame: Frame):
+            frame_desc = FrameDesc(frame.code_position_short, frame.function)
+            frame_desc_to_self_time[frame_desc] = (
+                frame_desc_to_self_time.get(frame_desc, 0) + frame.total_self_time
+            )
+            frame_desc_to_frame[frame_desc] = frame
+            frame_desc_to_ncalls[frame_desc] = frame_desc_to_ncalls.get(frame_desc, 0) + 1
+
+            for child in frame.children:
+                walk(child)
+
+        frame_desc_to_self_time: Dict[FrameDesc, float] = {}
+        frame_desc_to_frame: Dict[FrameDesc, Frame] = {}
+        frame_desc_to_ncalls: Dict[FrameDesc, int] = {}
+
+        walk(frame)
+
+        cost_list: List[Tuple[FrameDesc, float]] = sorted(
+            frame_desc_to_self_time.items(), key=(lambda item: item[1]), reverse=True
+        )
+
+        res = ""
+
+        for frame_desc, self_time in cost_list:
+            if self.time == "percent_of_total":
+                val = self_time / frame.time * 100
+                unit = "%"
+            else:
+                val = self_time
+                unit = "s"
+
+            color = self._ansi_color_for_time(self_time)
+
+            res += "{color}{val:.3f}{unit}{c.end} {name_color}{function}{c.end}  {c.faint}{code_position}{c.end} {ncalls}\n".format(
+                color=color,
+                val=val,
+                unit=unit,
+                c=self.colors,
+                name_color=self._ansi_color_for_name(frame_desc_to_frame[frame_desc]),
+                function=frame_desc.function,
+                code_position=frame_desc.code_position_short,
+                ncalls=frame_desc_to_ncalls[frame_desc],
+            )
```

Comment:
I'm wondering if there's any potential for code reuse of this bit. perhaps we can factor out a bit of  `render_frame` above - as an example, I note that this code won't print class names of functions... i think that there might be a potential for a `frame_description(self, frame: Frame, time: float) -> str` that captures all the text display stuff

Reply:
That's a good idea, thanks! I added `frame_description()` in [1171f3e](https://github.com/joerick/pyinstrument/pull/240/commits/1171f3e8d8f8e7b503e34d78d9a7e2ec19a898d7)